File: yumi/sdk/typescript/src/schema.ts
import type { ToolParameter, RegisterOptions } from "./types";

export function buildToolSchema(opts: RegisterOptions): Record<string, unknown> {
  // Map the `mode` API onto the existing wire flags (one mode per tool).
  let alwaysInclude = opts.alwaysInclude ?? false;
  let proactiveContext = opts.proactiveContext ?? false;
  let proactiveContextArgs = opts.proactiveContextArgs;
  let proactiveContextDescription = opts.proactiveContextDescription;
  const mode = opts.mode ?? "dynamic";
  if (mode === "pinned") {
    alwaysInclude = true;
  } else if (mode === "autorun") {
    proactiveContext = true;
    if (opts.contextArgs != null) proactiveContextArgs = opts.contextArgs;
    if (opts.contextLabel != null) proactiveContextDescription = opts.contextLabel;
  } else if (mode !== "dynamic") {
    throw new Error(
      `mode must be 'dynamic', 'pinned', or 'autorun'; got ${JSON.stringify(mode)}`
    );
  }

  const properties: Record<string, Record<string, unknown>> = {};
  const required: string[] = [];

  for (const param of opts.parameters ?? []) {
    properties[param.name] = {
      type: param.type,
      description: param.description,
    };
    if (param.required !== false) {
      required.push(param.name);
    }
  }

  const schema: Record<string, unknown> = {
    type: "function",
    function: {
      name: opts.name,
      description: opts.description,
      parameters: {
        type: "object",
        properties,
        required,
      },
    },
  };

  if (opts.timeout != null) {
    schema.timeout = opts.timeout;
  }
  if (opts.requireConfirmation) {
    schema.require_confirmation = true;
  }
  if (alwaysInclude) {
    schema.always_include = true;
  }
  if (opts.allowProactive) {
    schema.allow_proactive = true;
  }
  if (proactiveContext) {
    schema.proactive_context = true;
  }
  if (proactiveContextArgs != null) {
    schema.proactive_context_args = proactiveContextArgs;
  }
  if (proactiveContextDescription) {
    schema.proactive_context_description = proactiveContextDescription;
  }

  return schema;
}

File: yumi/sdk/go/schema.go
package yumi_sdk

import "fmt"

// BuildToolSchema creates the JSON schema object for a registered tool,
// matching the wire format expected by the Yumi server.
//
// It panics if opts.Mode is set to anything other than "dynamic", "pinned",
// or "autorun" (an invalid mode is a programming error, surfaced at register time).
func BuildToolSchema(opts RegisterOptions) map[string]interface{} {
	// Map the Mode API onto the existing wire flags (one mode per tool).
	alwaysInclude := opts.AlwaysInclude
	proactiveContext := opts.ProactiveContext
	proactiveContextArgs := opts.ProactiveContextArgs
	proactiveContextDescription := opts.ProactiveContextDescription
	switch opts.Mode {
	case "", "dynamic":
		// no change
	case "pinned":
		alwaysInclude = true
	case "autorun":
		proactiveContext = true
		if opts.ContextArgs != nil {
			proactiveContextArgs = opts.ContextArgs
		}
		if opts.ContextLabel != "" {
			proactiveContextDescription = opts.ContextLabel
		}
	default:
		panic(fmt.Sprintf("mode must be 'dynamic', 'pinned', or 'autorun'; got %q", opts.Mode))
	}

	properties := make(map[string]interface{})
	required := make([]string, 0)

	for _, p := range opts.Parameters {
		prop := map[string]interface{}{
			"type":        p.Type,
			"description": p.Description,
		}
		properties[p.Name] = prop

		isRequired := true
		if p.Required != nil {
			isRequired = *p.Required
		}
		if isRequired {
			required = append(required, p.Name)
		}
	}

	schema := map[string]interface{}{
		"type": "function",
		"function": map[string]interface{}{
			"name":        opts.Name,
			"description": opts.Description,
			"parameters": map[string]interface{}{
				"type":       "object",
				"properties": properties,
				"required":   required,
			},
		},
	}

	if opts.Timeout != nil {
		schema["timeout"] = *opts.Timeout
	}
	if opts.RequireConfirmation {
		schema["require_confirmation"] = true
	}
	if alwaysInclude {
		schema["always_include"] = true
	}
	if opts.AllowProactive {
		schema["allow_proactive"] = true
	}
	if proactiveContext {
		schema["proactive_context"] = true
	}
	if proactiveContextArgs != nil {
		schema["proactive_context_args"] = proactiveContextArgs
	}
	if proactiveContextDescription != "" {
		schema["proactive_context_description"] = proactiveContextDescription
	}

	return schema
}
